### PR TITLE
Implement WithExplicitBucketBoundaries option in the metric SDK

### PR DESCRIPTION
**Pull request by dashpole**
Fixes https://github.com/open-telemetry/opentelemetry-go/issues/4094

I looked at both #4341 and #4340 when implementing this, but mostly relied on #4340.  I omitted changes to the existing lookup() and aggs() functions (made in #4340) to minimize the diff.  If we still want those changes, I'd prefer refactoring as a follow-up.

The precedence of bucket boundaries is implemented as follows:

1. Boundaries specified in the View
2. Boundaries specified on the instrument
3. Boundaries provided by the reader, or the default boundaries if the reader uses the DefaultAggregation

### CODE REVIEW

File: sdk/metric/meter.go [hunk 1]
```diff
@@ -456,12 +454,36 @@ func (p int64InstProvider) aggs(kind InstrumentKind, name, desc, u string) ([]ag
 	return p.int64Resolver.Aggregators(inst)
 }
 
+func (p int64InstProvider) histogramAggs(name string, cfg metric.Int64HistogramConfig) ([]aggregate.Measure[int64], error) {
+	boundaries := cfg.ExplicitBucketBoundaries()
+	aggError := AggregationExplicitBucketHistogram{Boundaries: boundaries}.err()
+	if aggError != nil {
+		// if boundaries are invalid, ignore them
```
Comment (MrAlias) [inline sdk/metric/meter.go:461]: nit

```suggestion
		// If boundaries are invalid, ignore them.
```

File: sdk/metric/meter.go [hunk 2]
```diff
@@ -476,12 +498,36 @@ func (p float64InstProvider) aggs(kind InstrumentKind, name, desc, u string) ([]
 	return p.float64Resolver.Aggregators(inst)
 }
 
+func (p float64InstProvider) histogramAggs(name string, cfg metric.Float64HistogramConfig) ([]aggregate.Measure[float64], error) {
+	boundaries := cfg.ExplicitBucketBoundaries()
+	aggError := AggregationExplicitBucketHistogram{Boundaries: boundaries}.err()
+	if aggError != nil {
+		// if boundaries are invalid, ignore them
```
Comment (MrAlias) [inline sdk/metric/meter.go:505]: nit

```suggestion
		// If boundaries are invalid, ignore them.
```

**Review by MrAlias — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - Add `Version` function in `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc`. (#4660)
 - Add `Version` function in `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp`. (#4660)
 - Add Summary, SummaryDataPoint, and QuantileValue to `go.opentelemetry.io/sdk/metric/metricdata`. (#4622)
+- Add support for WithExplicitBucketBoundaries in `go.opentelemetry.io/otel/sdk/metric` (#4605)
```
Comment (pellared) [inline CHANGELOG.md:23]:
```suggestion
- Add support for `WithExplicitBucketBoundaries` in `go.opentelemetry.io/otel/sdk/metric`. (#4605)
```

**Review by pellared — APPROVED** [review]

**Review by MadVikingGod — APPROVED** [review]
I'm ok with this overall.

How much do you think would need to change if there was a `WithExponential()` option?

**Comment by dashpole** [comment]
> How much do you think would need to change if there was a WithExponential() option?

You would basically just change HistogramAggregators.  It would probably take the full config instead of buckets as a separate argument.  But overall, not much change I think.

**Pull request merged**